### PR TITLE
Added dynamic authentication carbon feature and updated pom

### PR DESCRIPTION
**Pull request by milanperera**

### CODE REVIEW

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderService.java [hunk 1]
```diff
@@ -79,4 +79,13 @@
     List<Device> getDevicesByName(String deviceName) throws DeviceManagementException;
 
     void updateDeviceEnrolmentInfo(Device device, EnrolmentInfo.Status active) throws DeviceManagementException;
+
+    /**
+     * This method is used to retrieve list of devices based on the device status
```
Comment (DilanUA) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderService.java:84]: need full-stop for the comment.

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java [hunk 2]
```diff
@@ -716,6 +716,45 @@ public void unregisterDeviceManagementService(DeviceManagementService deviceMana
...
 
+
+    public List<Device> getDevicesByStatus(EnrolmentInfo.Status status) throws DeviceManagementException {
+        List<Device> devices = new ArrayList<Device>();
+        List<Device> allDevices;
+        try {
+            DeviceManagementDAOFactory.getConnection();
+            int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+            allDevices = deviceDAO.getDevicesByStatus(status, tenantId);
+
```
Comment (DilanUA) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:727]: unnecessary empty line here...

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java [hunk 3]
```diff
@@ -716,6 +716,45 @@ public void unregisterDeviceManagementService(DeviceManagementService deviceMana
...
+        } catch (DeviceManagementDAOException e) {
+            String errorMsg = "Error occurred while fetching the list of devices that matches to status: '"
+                              + status + "'";
+            log.error(errorMsg, e);
+            throw new DeviceManagementException(errorMsg, e);
+        } finally {
+
```
Comment (DilanUA) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:734]: unnecessary empty line here...

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java [hunk 4]
```diff
@@ -716,6 +716,45 @@ public void unregisterDeviceManagementService(DeviceManagementService deviceMana
...
+                DeviceManagementDAOFactory.closeConnection();
+            } catch (DeviceManagementDAOException e) {
+                log.warn("Error occurred while closing the connection", e);
+            }
+        }
+
+        for (Device device : allDevices) {
+            Device dmsDevice =
+                    this.getPluginRepository().getDeviceManagementService(
+                            device.getType()).getDeviceManager().getDevice(
```
Comment (DilanUA) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:745]: You can use private method getDeviceManager() straightaway here to get an instance of corresponding device manager impl. This also checks for the null-able scenario of this.getPluginRepository().getDeviceManagementService()

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java [hunk 5]
```diff
@@ -716,6 +716,45 @@ public void unregisterDeviceManagementService(DeviceManagementService deviceMana
         }
     }
 
+
+    public List<Device> getDevicesByStatus(EnrolmentInfo.Status status) throws DeviceManagementException {
+        List<Device> devices = new ArrayList<Device>();
+        List<Device> allDevices;
+        try {
+            DeviceManagementDAOFactory.getConnection();
```
Comment (DilanUA) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:724]: shouldn't this be DeviceManagementDAOFactory.openConnection();

File: features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+    <description>This feature contains oauth functionality
+    </description>
+
+    <!--<dependencies>-->
+
+        <!--<dependency>-->
+            <!--<groupId>org.wso2.mdm</groupId>-->
+            <!--<artifactId>dynamic-client-manager</artifactId>-->
+        <!--</dependency>-->
+    <!--</dependencies>-->
```
Comment (DilanUA) [inline features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml:46]: You can remove these commented lines if not using...

File: features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+                                    <destFileName>dynamic-client-manager.war</destFileName>
+                                </artifactItem>
+                            </artifactItems>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+
+
+
```
Comment (DilanUA) [inline features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml:105]: unnecessary empty lines here...

File: features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+                                </importFeatureDef>
+                            </importFeatures>
+                        </configuration>
+                    </execution>
+                </executions>
+            </plugin>
+
+
+
+
```
Comment (DilanUA) [inline features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml:151]: unnecessary empty lines here...

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderService.java (hunk 1 above)
Comment (milanperera) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderService.java:84]: Added in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java (hunk 5 above)
Comment (milanperera) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:724]: Yes. It should be. This is fixed in commit [1].
[1] https://github.com/wso2/carbon-device-mgt/commit/00807e8295c06fc4241caebcc514319ae0d88f3b

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java (hunk 2 above)
Comment (milanperera) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:727]: Fixed in commit [1].
[1] https://github.com/wso2/carbon-device-mgt/commit/00807e8295c06fc4241caebcc514319ae0d88f3b

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java (hunk 3 above)
Comment (milanperera) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:734]: Fixed in commit [1].
[1] https://github.com/wso2/carbon-device-mgt/commit/00807e8295c06fc4241caebcc514319ae0d88f3b

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java (hunk 4 above)
Comment (milanperera) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:745]: Changed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27/

File: features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml (hunk 8 above)
Comment (milanperera) [inline features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml:151]: Removed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml (hunk 7 above)
Comment (milanperera) [inline features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml:105]: Removed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml (hunk 6 above)
Comment (milanperera) [inline features/oauth-extensions/org.wso2.carbon.oauth.extensions.server.feature/pom.xml:46]: Removed in PR [1].
[1] https://github.com/wso2/carbon-device-mgt/pull/27

File: components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java [hunk 9]
```diff
@@ -716,6 +716,45 @@ public void unregisterDeviceManagementService(DeviceManagementService deviceMana
...
+            int tenantId = CarbonContext.getThreadLocalCarbonContext().getTenantId();
+            allDevices = deviceDAO.getDevicesByStatus(status, tenantId);
+
+        } catch (DeviceManagementDAOException e) {
+            String errorMsg = "Error occurred while fetching the list of devices that matches to status: '"
+                              + status + "'";
+            log.error(errorMsg, e);
```
Comment (prabathabey) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:731]: Avoid log and through unless specifically needed.

Comment (milanperera) [inline components/device-mgt/org.wso2.carbon.device.mgt.core/src/main/java/org/wso2/carbon/device/mgt/core/service/DeviceManagementProviderServiceImpl.java:731]: Removed in commit [https://github.com/wso2/carbon-device-mgt/commit/00807e8295c06fc4241caebcc514319ae0d88f3b]

**Pull request merged**